### PR TITLE
testutil/promrated: implement rated fetching and reporting

### PR DESCRIPTION
**Pull request by LukeHackett12**
Adds fetching of rated effectiveness and making these metrics available to prometheus.

category: misc
ticket: https://github.com/ObolNetwork/charon/issues/1540

### CODE REVIEW

File: testutil/promrated/metrics.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package promrated
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+type watcherMetrics struct {
```
Comment (corverroos) [inline testutil/promrated/metrics.go:23]: please follow charon metrics patterns:
 - use charon's promauto 
 - use global counter/gauge variables
 - at startup, get them via promauto.NewRegistry
 - serve that registry via the monitoring api

File: testutil/promrated/metrics.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	"github.com/obolnetwork/charon/app/promauto"
+)
+
+var (
+	uptime = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "rated",
+		Subsystem: "sentinel",
+		Name:      "validation_key_uptime",
+		Help:      "Uptime of a validation key.",
+	}, []string{"pubkey_full", "cluster_name", "cluster_hash", "cluster_network"})
```
Comment (corverroos) [inline testutil/promrated/metrics.go:30]: can make labels a variable since it is duplicated everywhere

File: testutil/promrated/promrated.go [hunk 3]
```diff
@@ -40,9 +43,16 @@ func Run(ctx context.Context, config Config) error {
 		z.Str("monitoring_addr", config.MonitoringAddr),
 	)
 
+	labels := map[string]string{}
+
+	promRegistry, err := promauto.NewRegistry(labels)
```
Comment (corverroos) [inline testutil/promrated/promrated.go:48]: can just use `nil` I think

File: testutil/promrated/promrated.go [hunk 4]
```diff
@@ -79,5 +90,25 @@ func reportMetrics(ctx context.Context, config Config) {
...
 		)
+
+		stats, err := getValidationStatistics(ctx, config.RatedEndpoint, validator)
+		if err != nil {
+			log.Error(ctx, "Getting validator statistics", err, z.Str("validator", validator.PubKey))
+			continue
+		}
+
+		clusterLabels := prometheus.Labels{
+			"pubkey_full":     validator.ClusterName,
```
Comment (corverroos) [inline testutil/promrated/promrated.go:101]: typo

File: testutil/promrated/rated.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+func getValidationStatistics(ctx context.Context, ratedEndpoint string, validator validator) (*validatorEffectivenessData, error) {
+	url, err := url.Parse(ratedEndpoint)
+	if err != nil {
+		return nil, errors.Wrap(err, "parse rated endpoint")
+	}
+
+	url.Path = fmt.Sprintf("/v0/eth/validators/%s/effectiveness", validator.PubKey)
+
+	query := url.Query()
+	query.Add("size", "1")
```
Comment (corverroos) [inline testutil/promrated/rated.go:52]: what does this do btw?

File: testutil/promrated/rated.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+			return body, nil
+		}
+
+		body, err := bodyFunc()
+		if err != nil {
+			return nil, err
+		}
+
+		if res.StatusCode == 429 {
+			sleepFor := expbackoff.Backoff(expbackoff.DefaultConfig, r)
```
Comment (corverroos) [inline testutil/promrated/rated.go:94]: suggest using `expbackoff.New`

File: testutil/promrated/rated.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+			clusterNetwork = "prater"
+		}
+
+		req.Header.Add("X-Rated-Network", clusterNetwork)
+		res, err := client.Do(req)
+		if err != nil {
+			return nil, errors.Wrap(err, "requesting rated matrics")
+		}
+
+		bodyFunc := func() ([]byte, error) {
```
Comment (corverroos) [inline testutil/promrated/rated.go:77]: no need for an abstraction if only called once

File: testutil/promrated/rated.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+	query := url.Query()
+	query.Add("size", "1")
+	url.RawQuery = query.Encode()
+
+	maxRetries := 10
+
+	client := new(http.Client)
+
+	for r := 0; r <= maxRetries; r++ {
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, url.String(), nil)
```
Comment (corverroos) [inline testutil/promrated/rated.go:60]: always define timeouts when doing IO: https://blog.cloudflare.com/the-complete-guide-to-golang-net-http-timeouts/

File: testutil/promrated/rated.go [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+		}
+
+		body, err := bodyFunc()
+		if err != nil {
+			return nil, err
+		}
+
+		if res.StatusCode == 429 {
+			sleepFor := expbackoff.Backoff(expbackoff.DefaultConfig, r)
+			log.Info(ctx, fmt.Sprintln("Rate limit exceeded. Retrying after .", sleepFor))
```
Comment (corverroos) [inline testutil/promrated/rated.go:95]: never use `fmt.SprintX` in logs or errors, use structured fields: `z.Int("attempt", r)`

File: testutil/promrated/rated.go [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+
+			continue
+		} else if res.StatusCode/100 != 2 {
+			return nil, errors.New("not ok http response", z.Str("body", string(body)))
+		}
+
+		return parseMetrics(body)
+	}
+
+	return nil, errors.New("fetching Validator data", z.Int("retries", maxRetries))
```
Comment (corverroos) [inline testutil/promrated/rated.go:106]: `return nil, errors.New("max retries exceeded fetching validator data", z.Int("max", maxRetries))`

File: testutil/promrated/rated.go [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+// parseMetrics reads rated response and returns the validator effectiveness data
+func parseMetrics(body []byte) (*validatorEffectivenessData, error) {
```
Comment (corverroos) [inline testutil/promrated/rated.go:110]: avoid pointers

File: testutil/promrated/rated.go [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, url.String(), nil)
+		if err != nil {
+			return nil, errors.Wrap(err, "new rated request")
+		}
+
+		//Workaround for rated api not recognising goerli
+		clusterNetwork := validator.ClusterNetwork
+		if clusterNetwork == "goerli" {
+			clusterNetwork = "prater"
+		}
```
Comment (corverroos) [inline testutil/promrated/rated.go:69]: move this out of the loop since it is constant.

File: testutil/promrated/rated.go [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+	Uptime                 float64 `json:"uptime"`
+	AvgCorrectness         float64 `json:"avgCorrectness"`
+	AvgInclusionDelay      float64 `json:"avgInclusionDelay"`
+	AttesterEffectiveness  float64 `json:"attesterEffectiveness"`
+	ProposerEffectiveness  float64 `json:"proposerEffectiveness"`
+	ValidatorEffectiveness float64 `json:"validatorEffectiveness"`
+}
+
+// getValidationStatistics queries rated for a pubkey and returns rated data about the pubkey
+func getValidationStatistics(ctx context.Context, ratedEndpoint string, validator validator) (*validatorEffectivenessData, error) {
```
Comment (corverroos) [inline testutil/promrated/rated.go:43]: maybe add a link to the ratedAPI spec

File: testutil/promrated/rated.go (hunk 5 above)
Comment (corverroos) [reply testutil/promrated/rated.go:52]: add comment what this means maybe

File: testutil/promrated/rated.go (hunk 7 above)
Comment (LukeHackett12) [reply testutil/promrated/rated.go:77]: It is in a loop and the linter was complaining about deferring in a loop, can move this function creation out though as it isn't necessary to do every time

File: testutil/promrated/rated.go [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+
+			return body, nil
+		}
+
+		body, err := bodyFunc()
+		if err != nil {
+			return nil, err
+		}
+
+		if res.StatusCode == 429 {
```
Comment (corverroos) [inline testutil/promrated/rated.go:93]: http.StatusTooManyRequests

File: testutil/promrated/rated.go (hunk 5 above)
Comment (LukeHackett12) [reply testutil/promrated/rated.go:54]: added comment, it means it is just fetching the most recent day of data

File: testutil/promrated/metrics.go [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+
+package promrated
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/obolnetwork/charon/app/promauto"
+)
+
+// metrics.
```
Comment (corverroos) [inline testutil/promrated/metrics.go:24]: can remove this comment, doesn't add much value

File: testutil/promrated/rated.go [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+// See https://api.rated.network/docs#/default/get_effectiveness_v0_eth_validators__validator_index_or_pubkey__effectiveness_get
+func getValidationStatistics(ctx context.Context, ratedEndpoint string, validator validator) (validatorEffectivenessData, error) {
+	url, err := url.Parse(ratedEndpoint)
+	if err != nil {
+		return validatorEffectivenessData{}, errors.Wrap(err, "parse rated endpoint")
+	}
+
+	url.Path = fmt.Sprintf("/v0/eth/validators/%s/effectiveness", validator.PubKey)
+
+	//Adding size=1 will get only the latest day of data
```
Comment (corverroos) [inline testutil/promrated/rated.go:52]: space after `//` before comment starts

File: testutil/promrated/rated.go [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+			return validatorEffectivenessData{}, errors.Wrap(err, "requesting rated matrics")
+		}
+
+		body, err := extractBody(res)
+		if err != nil {
+			return validatorEffectivenessData{}, err
+		}
+
+		if res.StatusCode == http.StatusTooManyRequests {
+			log.Info(ctx, "Rate limit exceeded.", z.Int("retry", r))
```
Comment (corverroos) [inline testutil/promrated/rated.go:89]: no fullstops in logs

**Review by corverroos — APPROVED** [review]

File: testutil/promrated/promrated.go [hunk 18]
```diff
@@ -79,5 +88,25 @@ func reportMetrics(ctx context.Context, config Config) {
...
+		clusterLabels := prometheus.Labels{
+			"pubkey_full":     validator.PubKey,
+			"cluster_name":    validator.ClusterName,
+			"cluster_hash":    validator.ClusterHash,
+			"cluster_network": validator.ClusterNetwork,
+		}
+
+		uptime.With(clusterLabels).Set(stats.Uptime)
+		correctness.With(clusterLabels).Set(stats.AvgCorrectness)
+		inclusionDelay.With(clusterLabels).Set(stats.AvgCorrectness)
```
Comment (corverroos) [inline testutil/promrated/promrated.go:107]: typo

**Pull request merged**